qiskit/circuit/quantumcircuitdata.py

```diff
@@ -59,52 +133,71 @@ def __setitem__(self, key, value):
+        try:
+            # To consolidate validation to __setitem__, insert None and overwrite.
+            self._list.insert(index, None)
+            self[index] = value
+        except CircuitError as err:
+            del self._list[index]
+            raise err
 
-    def __len__(self):
-        return len(self._circuit._data)
+    def __delitem__(self, index):
+        del self._list[index]
 
-    def __cast(self, other):
-        return other._circuit._data if isinstance(other, QuantumCircuitData) else other
 
-    def __repr__(self):
-        return repr(self._circuit._data)
-
-    def __lt__(self, other):
-        return self._circuit._data < self.__cast(other)
+class QuantumCircuitQregs(ValidatedMutableSequence):
+    """A wrapper class for the purposes of validating modifications to
+    QuantumCircuit.qregs while maintaining the interface of a python list."""
 
-    def __le__(self, other):
-        return self._circuit._data <= self.__cast(other)
+    def __init__(self, circuit):
+        self._circuit = circuit
+        super().__init__(circuit._qregs)
 
-    def __eq__(self, other):
-        return self._circuit._data == self.__cast(other)
+    def __setitem__(self, key, value):
+        self._circuit.add_register(value)
+        self._list[key] = value
+        del self._list[-1]
 
-    def __gt__(self, other):
-        return self._circuit._data > self.__cast(other)
+    def insert(self, index, value):
+        self._circuit.add_register(value)
+        self._list.insert(index, self._list[-1])
+        del self._list[-1]
 
-    def __ge__(self, other):
-        return self._circuit._data >= self.__cast(other)
+    def __delitem__(self, index):
+        del self._list[index]
 
-    def __add__(self, other):
-        return self._circuit._data + self.__cast(other)
+    def reverse(self):
+        # Override MutableSequence.reverse which uses repeated pairwise
+        # calls to __setitem__ that do not maintain uniqueness of registers
+        # in _list throughout, and so raises a "CircuitError: register name
+        # ... already exists"
+        self._list.reverse()
 
-    def __radd__(self, other):
-        return self.__cast(other) + self._circuit._data
 
-    def __mul__(self, n):
-        return self._circuit._data * n
+class QuantumCircuitCregs(ValidatedMutableSequence):
+    """A wrapper class for the purposes of validating modifications to
+    QuantumCircuit.cregs while maintaining the interface of a python list."""
 
-    def __rmul__(self, n):
-        return n * self._circuit._data
+    def __init__(self, circuit):
+        self._circuit = circuit
+        super().__init__(circuit._cregs)
 
-    def sort(self, *args, **kwargs):
-        """In-place stable sort. Accepts arguments of list.sort."""
-        self._circuit._data.sort(*args, **kwargs)
+    def __setitem__(self, key, value):
+        self._circuit.add_register(value)
+        self._list[key] = value
+        del self._list[-1]
```

Comment:
The circular logic here is difficult to follow for me.  As I understand it:
1. `self._circuit.add_register(value)` ends up calling `self.append(value)`
2. now that the bit references are in place within `self._circuit`, we set the actual element of our list to the correct value
3. we delete the extra data that the unwanted `self.append(value)` added

My worries here are:
1. it feels like we actually just want something like `self._circuit._add_internal_bit_references(value)`, since in the end, _this_ class is responsible for the append.
2. this won't work if `key` is negative, or a `slice`, both of which are supported by list syntax.  For example, `self[-1] = QuantumRegister(2)` would append the quantum register to the end of the list, overwrite the same list location with the same data, and then delete it off the end.  A slice would cause a list to get nested in the middle.

I think the solution to this involves changing the call to `self._circuit.add_register`, so that the method by which we register bits in the circuit does not affect `self`, and from there we can delegate to `self._list.__setitem__` entirely.  This would probably involve splitting out the "add to list of registers" logic from the "add references to all bits logic" inside `QubitCircuit.add_register`, so we can only call the latter.